examples/imgui_impl_x11.cpp

```diff
@@ -0,0 +1,171 @@
+    io.KeyMap[ImGuiKey_Space] = XK_space;
+    io.KeyMap[ImGuiKey_Enter] = XK_Return - 0xFF00;
+    io.KeyMap[ImGuiKey_Escape] = XK_Escape - 0xFF00;
+    io.KeyMap[ImGuiKey_KeyPadEnter] = XK_KP_Enter - 0xFF00;
+    io.KeyMap[ImGuiKey_A] = 'A';
+    io.KeyMap[ImGuiKey_C] = 'C';
+    io.KeyMap[ImGuiKey_V] = 'V';
+    io.KeyMap[ImGuiKey_X] = 'X';
+    io.KeyMap[ImGuiKey_Y] = 'Y';
+    io.KeyMap[ImGuiKey_Z] = 'Z';
+
+    return true;
+}
+
+void    ImGui_ImplX11_Shutdown()
+{
+    xcb_key_symbols_free(g_KeySyms);
+    xcb_flush(g_Connection);
+    xcb_disconnect(g_Connection);
+}
+
+void    ImGui_ImplX11_NewFrame()
+{
+    ImGuiIO& io = ImGui::GetIO();
+    IM_ASSERT(io.Fonts->IsBuilt() && "Font atlas not built! It is generally built by the renderer back-end. Missing call to renderer _NewFrame() function? e.g. ImGui_ImplOpenGL3_NewFrame().");
+
+    // Setup time step
+    clock_gettime(CLOCK_MONOTONIC_RAW, &g_CurrentTime);
+    io.DeltaTime = (g_CurrentTime.tv_sec - g_LastTime.tv_sec) +
+        ((g_CurrentTime.tv_nsec - g_LastTime.tv_nsec) / 1000000000.0f);
+    g_LastTime = g_CurrentTime;
+}
+
+// X11 xcb message handler (process X11 mouse/keyboard inputs, etc.)
+// Call from your application's message handler. Returns true if
+// ImGui processed the event. You can use this for checking if you
+// need further processing for the event.
+bool ImGui_ImplX11_Event(xcb_generic_event_t *event)
+{
+    if (ImGui::GetCurrentContext() == NULL)
+        return false;
+
+    ImGuiIO& io = ImGui::GetIO();
+    switch (event->response_type & ~0x80) {
+    case XCB_MOTION_NOTIFY: {
+        xcb_motion_notify_event_t *e = (xcb_motion_notify_event_t*)event;
+        io.MousePos = ImVec2(e->event_x, e->event_y);
+        return true;
+    }
+    case XCB_KEY_PRESS: {
+        xcb_key_press_event_t *e = (xcb_key_press_event_t *)event;
+        xcb_keysym_t k = xcb_key_press_lookup_keysym(g_KeySyms, e, 0);
+        if(k < 0xFF) { // latin-1 range
+            io.AddInputCharacter(k);
+            io.KeysDown[k] = 1;
+        }
+        else if(k >= 0x1000100 && k <= 0x110ffff) // utf range
+            io.AddInputCharacterUTF16(k);
+        else
+            io.KeysDown[k - 0xFF00] = 1;
+
+        return true;
+    }
+    case XCB_KEY_RELEASE: {
+        xcb_key_press_event_t *e = (xcb_key_press_event_t *)event;
+        xcb_keysym_t k = xcb_key_press_lookup_keysym(g_KeySyms, e, 0);
+        if(k < 0xFF)
+            io.KeysDown[k] = 0;
+        else
+            io.KeysDown[k - 0xFF00] = 0;
+
+        return true;
+    }
+    case XCB_BUTTON_PRESS: {
+        xcb_button_press_event_t *e = (xcb_button_press_event_t *)event;
+        // Get exact coords of the event. It may be separate from the mouse cursor.
+        // e.g. touch input
+        io.MousePos = ImVec2(e->event_x, e->event_y);
+        // X decided button 4 is mwheel up and 5 is mwheel down
+        if(e->detail == 4)
+            io.MouseWheel += 1.0;
```

Comment:
Win32 has GET_WHEEL_DELTA_WPARAM() and `WHEEL_DELTA` macros for figuring out how 'far' the 'mouse wheel' (potentially touch/trackpad scroll) went. I couldn't find any such thing in X11 proper, that's usually handled by the window manager in different and of course incompatible ways. This works fine with touchpad and mwheel events, and a literal +/-1 felt fine.